test/Chainweb/Test/Pact/RemotePactTest.hs

```diff
@@ -273,6 +277,93 @@ localChainDataTest iot nio = do
           assert' name value = assertEqual name (M.lookup  (FieldKey (T.pack name)) m) (Just value)
     expectedResult _ = assertFailure "Didn't get back an object map!"
 
+localPreflightSimTest :: IO (Time Micros) -> IO ChainwebNetwork -> TestTree
+localPreflightSimTest iot nio = testCaseSteps "local preflight sim test" $ \step -> do
+    cenv <- _getServiceClientEnv <$> nio
+    mv <- newMVar (0 :: Int)
+    sid <- mkChainId v maxBound (0 :: Int)
+    let sigs = [mkSigner' sender00 []]
+
+    step "Execute preflight /local tx - preflight known /send success"
+    let psid = Pact.ChainId $ chainIdToText sid
+    psigs <- testKeyPairs sender00 Nothing
+    cmd0 <- mkRawTx mv psid psigs
+    runLocalPreflightClient sid cenv cmd0 >>= \case
+      Left e -> assertFailure $ show e
+      Right{} -> pure ()
+
+    step "Execute preflight /local tx - unparseable chain id"
+    sigs0 <- testKeyPairs sender00 Nothing
+    cmd1 <- mkRawTx mv (Pact.ChainId "fail") sigs0
+    runClientFailureAssertion sid cenv cmd1 "Unparseable transaction chain id"
+
+    step "Execute preflight /local tx - chain id mismatch"
+    let fcid = unsafeChainId (-1)
+    cmd2 <- mkTx mv =<< mkCmdBuilder sigs v fcid 1000 gp
+    runClientFailureAssertion sid cenv cmd2 "Chain id mismatch"
+
+    step "Execute preflight /local tx - tx gas limit too high"
+    cmd3 <- mkTx mv =<< mkCmdBuilder sigs v sid 100000000000000 gp
+    runClientFailureAssertion sid cenv cmd3
+      "Transaction Gas limit exceeds block gas limit"
+
+    step "Execute preflight /local tx - tx gas price precision too high"
+    cmd4 <- mkTx mv =<< mkCmdBuilder sigs v sid 1000 0.00000000000000001
+    runClientFailureAssertion sid cenv cmd4
+      "Gas price decimal precision too high"
+
+    step "Execute preflight /local tx - network id mismatch"
+    cmd5 <- mkTx mv =<< mkCmdBuilder sigs Mainnet01 sid 1000 gp
+    runClientFailureAssertion sid cenv cmd5 "Network id mismatch"
+
+    step "Execute preflight /local tx - too many sigs"
+    let pcid = Pact.ChainId $ chainIdToText sid
+    sigs1' <- (>>= replicate 101) <$> testKeyPairs sender00 Nothing
```

Comment:
Comprehension-wise, this one is a bit nasty. I think you're using `<$>` to fmap into something that returns a list, and then you're using the list monad's bind operator to generate a list of lists? Is there some way to use a point to make this a bit more transparent to future editors?

Reply:
It's actually a bit of a terse hack - `testKeyPairs` always produces a singleton list, so i'm using the list monad instance to join a singleton list of lists, which is... a list. I can make this less opaque if you want.

Reply:
Updated to be more readable